a64 release build failures with gcc 10: zero-sized arrays

Xref Visual Studio hitting similar errors about zero-sized arrays at 
https://github.com/DynamoRIO/dynamorio/issues/1684#issuecomment-648931235

```
$ /usr/bin/aarch64-linux-gnu-gcc --version
aarch64-linux-gnu-gcc (Debian 10.2.0-9) 10.2.0
$ rm -rf *; cmake -GNinja -DCMAKE_TOOLCHAIN_FILE=../src/make/toolchain-arm64.cmake ../src; ninja dynamorio
[69/104] Building C object core/CMakeFiles/dynamorio.dir/arch/clean_call_opt_shared.c.o   
FAILED: core/CMakeFiles/dynamorio.dir/arch/clean_call_opt_shared.c.o 
/usr/bin/aarch64-linux-gnu-gcc -Ddynamorio_EXPORTS -I/home/bruening/dr/git/src/core/drlibc -I/home/bruening/dr/git/src/core/ir/aarch64 -I/home/bruening/dr/git/src/core/arch/aarch64 -I/home/bruening/dr/git/src/core/unix -I/home/bruening/dr/git/src/core/ir -I/home/bruening/dr/git/src/core/arch -I/home/bruening/dr/git/src/core/lib -I. -fno-strict-aliasing -fno-stack-protector -fvisibility=internal  -std=gnu99 -fno-unwind-tables -O3 -g3 -Wall -Werror -Wwrite-strings -Wno-unused-but-set-variable -Wno-stringop-truncation -Wno-format-truncation -Wno-stringop-overflow -fPIC -MD -MT core/CMakeFiles/dynamorio.dir/arch/clean_call_opt_shared.c.o -MF core/CMakeFiles/dynamorio.dir/arch/clean_call_opt_shared.c.o.d -o core/CMakeFiles/dynamorio.dir/arch/clean_call_opt_shared.c.o -c /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c                                                                                                                         
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c: In function ‘analyze_clean_call’:                                             
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:81:24: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                             
   81 |         ci->opmask_used[i] = true;                               
      |         ~~~~~~~~~~~~~~~^~~
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:1463:10: note: while referencing ‘opmask_used’                                                  
 1463 |     bool opmask_used[MCXT_NUM_OPMASK_SLOTS];                          
      |          ^~~~~~~~~~~                                                  
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:81:24: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                             
   81 |         ci->opmask_used[i] = true;
      |         ~~~~~~~~~~~~~~~^~~
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:1463:10: note: while referencing ‘opmask_used’                                                  
 1463 |     bool opmask_used[MCXT_NUM_OPMASK_SLOTS];
      |          ^~~~~~~~~~~                                                  
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:517:30: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                            
  517 |         if (info->opmask_used[i]) {                          
      |             ~~~~~~~~~~~~~~~~~^~~                                      
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:1463:10: note: while referencing ‘opmask_used’                                                  
 1463 |     bool opmask_used[MCXT_NUM_OPMASK_SLOTS];
      |          ^~~~~~~~~~~                                                  
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:523:29: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                            
  523 |             cci->opmask_skip[i] = true;                               
      |             ~~~~~~~~~~~~~~~~^~~                                       
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:412:10: note: while referencing ‘opmask_skip’                                                   
  412 |     bool opmask_skip[MCXT_NUM_OPMASK_SLOTS];                          
      |          ^~~~~~~~~~~
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:518:29: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                            
  518 |             cci->opmask_skip[i] = false;                              
      |             ~~~~~~~~~~~~~~~~^~~                                       
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:412:10: note: while referencing ‘opmask_skip’                                                   
  412 |     bool opmask_skip[MCXT_NUM_OPMASK_SLOTS];                          
      |          ^~~~~~~~~~~
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c: In function ‘clean_call_opt_init’:                                            
/home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:81:24: error: array subscript ‘(<unknown>) + -1’ is outside the bounds of an interior zero-length array ‘bool[0]’ {aka ‘_Bool[]’} [-Werror=zero-length-bounds]                                                                             
   81 |         ci->opmask_used[i] = true;                                    
      |         ~~~~~~~~~~~~~~~^~~
In file included from /home/bruening/dr/git/src/core/arch/clean_call_opt_shared.c:42:                                                      
/home/bruening/dr/git/src/core/arch/arch.h:1463:10: note: while referencing ‘opmask_used’                                                  
 1463 |     bool opmask_used[MCXT_NUM_OPMASK_SLOTS];                          
      |          ^~~~~~~~~~~   
/home/bruening/dr/git/src/core/arch/arch.h:1478:22: note: defined here ‘default_callee_info’                                               
 1478 | extern callee_info_t default_callee_info;                             
      |                      ^~~~~~~~~~~~~~~~~~~                              
cc1: all warnings being treated as errors   
```

--- a/core/arch/arch.h
+++ b/core/arch/arch.h
@@ -408,8 +408,10 @@ typedef struct _clean_call_info_t {
     bool skip_clear_flags;
     int num_simd_skip;
     bool simd_skip[MCXT_NUM_SIMD_SLOTS];
+#ifdef X86
     int num_opmask_skip;
     bool opmask_skip[MCXT_NUM_OPMASK_SLOTS];
+#endif
     uint num_regs_skip;
     bool reg_skip[DR_NUM_GPR_REGS];
     bool preserve_mcontext; /* even if skip reg save, preserve mcontext shape */
@@ -1447,20 +1449,22 @@ typedef struct _slot_t {
 
 /* data structure of clean call callee information. */
 typedef struct _callee_info_t {
-    bool bailout;        /* if we bail out on function analysis */
-    uint num_args;       /* number of args that will passed in */
-    int num_instrs;      /* total number of instructions of a function */
-    app_pc start;        /* entry point of a function  */
-    app_pc bwd_tgt;      /* earliest backward branch target */
-    app_pc fwd_tgt;      /* last forward branch target */
-    int num_simd_used;   /* number of SIMD registers (xmms) used by callee */
-    int num_opmask_used; /* number of mask registers used by callee */
+    bool bailout;      /* if we bail out on function analysis */
+    uint num_args;     /* number of args that will passed in */
+    int num_instrs;    /* total number of instructions of a function */
+    app_pc start;      /* entry point of a function  */
+    app_pc bwd_tgt;    /* earliest backward branch target */
+    app_pc fwd_tgt;    /* last forward branch target */
+    int num_simd_used; /* number of SIMD registers (xmms) used by callee */
     /* SIMD ([xyz]mm) registers usage. Part of the array might be left
      * uninitialized if proc_num_simd_registers() < MCXT_NUM_SIMD_SLOTS.
      */
     bool simd_used[MCXT_NUM_SIMD_SLOTS];
+#ifdef X86
+    int num_opmask_used; /* number of mask registers used by callee */
     /* AVX-512 mask register usage. */
     bool opmask_used[MCXT_NUM_OPMASK_SLOTS];
+#endif
     bool reg_used[DR_NUM_GPR_REGS];         /* general purpose registers usage */
     int num_callee_save_regs;               /* number of regs callee saved */
     bool callee_save_regs[DR_NUM_GPR_REGS]; /* callee-save registers */

--- a/core/arch/clean_call_opt_shared.c
+++ b/core/arch/clean_call_opt_shared.c
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2010-2019 Google, Inc.  All rights reserved.
+ * Copyright (c) 2010-2020 Google, Inc.  All rights reserved.
  * Copyright (c) 2016 ARM Limited. All rights reserved.
  * Copyright (c) 2010 Massachusetts Institute of Technology  All rights reserved.
  * Copyright (c) 2000-2010 VMware, Inc.  All rights reserved.
@@ -74,11 +74,13 @@ callee_info_init(callee_info_t *ci)
      */
     /* assuming all [xyz]mm registers are used */
     ci->num_simd_used = proc_num_simd_registers();
-    ci->num_opmask_used = proc_num_opmask_registers();
     for (i = 0; i < proc_num_simd_registers(); i++)
         ci->simd_used[i] = true;
+#    ifdef X86
+    ci->num_opmask_used = proc_num_opmask_registers();
     for (i = 0; i < proc_num_opmask_registers(); i++)
         ci->opmask_used[i] = true;
+#    endif
     for (i = 0; i < DR_NUM_GPR_REGS; i++)
         ci->reg_used[i] = true;
     ci->spill_reg = DR_REG_INVALID;
@@ -428,12 +430,14 @@ analyze_callee_inline(dcontext_t *dcontext, callee_info_t *ci)
             "CLEANCALL: callee " PFX " cannot be inlined: uses SIMD.\n", ci->start);
         opt_inline = false;
     }
+#    ifdef X86
     if (ci->num_opmask_used != 0) {
         LOG(THREAD, LOG_CLEANCALL, 1,
             "CLEANCALL: callee " PFX " cannot be inlined: uses mask register.\n",
             ci->start);
         opt_inline = false;
     }
+#    endif
     if (ci->tls_used) {
         LOG(THREAD, LOG_CLEANCALL, 1,
             "CLEANCALL: callee " PFX " cannot be inlined: accesses TLS.\n", ci->start);
@@ -513,6 +517,7 @@ analyze_clean_call_regs(dcontext_t *dcontext, clean_call_info_t *cci)
             cci->num_simd_skip++;
         }
     }
+#    ifdef X86
     for (i = 0; i < proc_num_opmask_registers(); i++) {
         if (info->opmask_used[i]) {
             cci->opmask_skip[i] = false;
@@ -524,6 +529,7 @@ analyze_clean_call_regs(dcontext_t *dcontext, clean_call_info_t *cci)
             cci->num_opmask_skip++;
         }
     }
+#    endif
     if (INTERNAL_OPTION(opt_cleancall) > 2 &&
         cci->num_simd_skip != proc_num_simd_registers())
         cci->should_align = false;
@@ -670,10 +676,12 @@ analyze_clean_call_inline(dcontext_t *dcontext, clean_call_info_t *cci)
         if (cci->num_simd_skip == proc_num_simd_registers()) {
             STATS_INC(cleancall_simd_skipped);
         }
+#    ifdef X86
         if (proc_num_opmask_registers() > 0 &&
             cci->num_opmask_skip == proc_num_opmask_registers()) {
             STATS_INC(cleancall_opmask_skipped);
         }
+#    endif
         if (cci->skip_save_flags) {
             STATS_INC(cleancall_aflags_save_skipped);
         }
@@ -755,7 +763,6 @@ analyze_clean_call(dcontext_t *dcontext, clean_call_info_t *cci, instr_t *where,
 #        define SIMD_SAVE_THRESHOLD 6
     /* Use out-of-line calls if more than 6 GP registers need to be saved. */
 #        define GPR_SAVE_THRESHOLD 6
-#        define OPMASK_SAVE_THRESHOLD 0
 #    endif
 
 #    if defined(X86) || defined(AARCH64)
@@ -766,8 +773,10 @@ analyze_clean_call(dcontext_t *dcontext, clean_call_info_t *cci, instr_t *where,
      * XXX: This should probably be in arch-specific clean_call_opt.c.
      */
     if ((proc_num_simd_registers() - cci->num_simd_skip) > SIMD_SAVE_THRESHOLD ||
-        (proc_num_opmask_registers() - cci->num_opmask_skip) > OPMASK_SAVE_THRESHOLD ||
-        (DR_NUM_GPR_REGS - cci->num_regs_skip) > GPR_SAVE_THRESHOLD || always_out_of_line)
+        IF_X86((proc_num_opmask_registers() - cci->num_opmask_skip) >
+                   OPMASK_SAVE_THRESHOLD ||)(DR_NUM_GPR_REGS - cci->num_regs_skip) >
+            GPR_SAVE_THRESHOLD ||
+        always_out_of_line)
         cci->out_of_line_swap = true;
 #    endif
 

--- a/core/arch/mangle_shared.c
+++ b/core/arch/mangle_shared.c
@@ -326,8 +326,8 @@ prepare_for_clean_call(dcontext_t *dcontext, clean_call_info_t *cci, instrlist_t
 #endif
         ASSERT((dstack_offs % get_ABI_stack_alignment()) == 0);
     }
-    ASSERT(cci->skip_save_flags || cci->num_simd_skip != 0 || cci->num_opmask_skip != 0 ||
-           cci->num_regs_skip != 0 ||
+    ASSERT(cci->skip_save_flags || cci->num_simd_skip != 0 ||
+           IF_X86(cci->num_opmask_skip != 0 ||) cci->num_regs_skip != 0 ||
            (int)dstack_offs ==
                (get_clean_call_switch_stack_size() + clean_call_beyond_mcontext()));
     return dstack_offs;

--- a/core/arch/mangle_shared.c
+++ b/core/arch/mangle_shared.c
@@ -125,8 +125,10 @@ insert_get_mcontext_base(dcontext_t *dcontext, instrlist_t *ilist, instr_t *wher
 bool
 clean_call_needs_simd(clean_call_info_t *cci)
 {
-    return (cci->preserve_mcontext || cci->num_simd_skip != proc_num_simd_registers() ||
-            cci->num_opmask_skip != proc_num_opmask_registers());
+    return (cci->preserve_mcontext ||
+            cci->num_simd_skip !=
+                proc_num_simd_registers()
+                    IF_X86(|| cci->num_opmask_skip != proc_num_opmask_registers()));
 }
 
 /* Number of extra slots in addition to register slots. */

--- a/core/lib/instrument.c
+++ b/core/lib/instrument.c
@@ -5628,12 +5628,14 @@ dr_insert_clean_call_ex_varg(void *drcontext, instrlist_t *ilist, instr_t *where
 #else
         /* all 8, 16 or 32 are scratch */
         cci.num_simd_skip = proc_num_simd_registers();
-        cci.num_opmask_skip = proc_num_opmask_registers();
 #endif
         for (i = 0; i < cci.num_simd_skip; i++)
             cci.simd_skip[i] = true;
+#ifdef X86
+        cci.num_opmask_skip = proc_num_opmask_registers();
         for (i = 0; i < cci.num_opmask_skip; i++)
             cci.opmask_skip[i] = true;
+#endif
             /* now remove those used for param/retval */
 #ifdef X64
         if (TEST(DR_CLEANCALL_NOSAVE_XMM_NONPARAM, save_flags)) {

--- a/core/lib/statsx.h
+++ b/core/lib/statsx.h
@@ -1239,7 +1239,9 @@ STATS_DEF("Clean Call analyzed", cleancall_analyzed)
 STATS_DEF("Clean Call inserted", cleancall_inserted)
 STATS_DEF("Clean Call inlined", cleancall_inlined)
 STATS_DEF("Clean Call [xyz]mm skipped", cleancall_simd_skipped)
+#ifdef X86
 STATS_DEF("Clean Call mask skipped", cleancall_opmask_skipped)
+#endif
 STATS_DEF("Clean Call aflags save skipped", cleancall_aflags_save_skipped)
 STATS_DEF("Clean Call aflags clear skipped", cleancall_aflags_clear_skipped)
 /* i#107 handle application using same segment register */

